book/tests/end_to_end.rs

```diff
@@ -0,0 +1,174 @@
+                    id: 2,
+                    first_name: "Maya".to_string(),
+                    last_name: "Alroy".to_string(),
+                    isni: None,
+                    orcid: None,
+                    oclc: None,
+                },
+            ],
+            subject_areas: vec![],
+        }
+    }
+}
+
+mod utils {
+    use std::io;
+    use std::net::SocketAddr;
+    use std::sync::Arc;
+
+    use diesel::prelude::*;
+    use tarpc::client;
+    use tokio_serde::formats::Json;
+    use uuid::Uuid;
+
+    use book::db::{self, DbConn, DbPool};
+    use book::rpc_server;
+    use book::service::BookServiceClient;
+
+    use super::*;
+
+    pub async fn test_setup() -> (DbConn, BookServiceClient) {
+        // database
+        let db_pool = db_setup();
+        let db_conn = db_pool.get().unwrap();
+        db_test_data(&db_conn).unwrap();
+        // rpc
+        let addr = spawn_rpc_server(db_pool).await;
+        let client = get_rpc_client(addr).await.unwrap();
+
+        (db_conn, client)
+    }
+
+    fn db_setup() -> DbPool {
+        // create dummy database
+        let tmp_conn = PgConnection::establish(DATABASE_URL).unwrap();
+        let db_name = Uuid::new_v4();
+        diesel::sql_query(&*format!("CREATE DATABASE \"{}\";", db_name))
+            .execute(&tmp_conn)
+            .expect("Couldn't create dummy database");
+
+        // connect to dummy database
+        let database_url = format!("{}/{}", DATABASE_URL, db_name);
+        let conn = db::new_db_pool(&*database_url);
+
+        // migrate dummy database
+        embed_migrations!();
+        helpers::db::run_migration(embedded_migrations::run, &conn);
+
+        conn
+    }
+
+    pub fn db_test_data(db_conn: &DbConn) -> QueryResult<()> {
+        let queries = vec![
+            "INSERT INTO persons (first_name, last_name) VALUES ('Markus', 'Merklinger');",
+            "INSERT INTO persons (first_name, last_name) VALUES ('Maya', 'Alroy');",
+            "INSERT INTO persons (first_name, last_name) VALUES ('Wail', 'Ahmad');",
+            "INSERT INTO publishers (name) VALUES ('Nostarch Press');",
+            "INSERT INTO publishers (name) VALUES ('MIT Press');",
+            "INSERT INTO books (code_identifier, isbn, release_date, title, subtitle, category_id, language_id, publisher_id) VALUES ('SE001', '1234567891011', '1900-01-01', 'Rust Book', 'Why you should learn it', 1, 1, 1);",
+            "INSERT INTO books (code_identifier, isbn, release_date, title, subtitle, category_id, language_id, publisher_id) VALUES ('SE002', '1234567891012', '2000-01-01', 'Python Book', DEFAULT, 1, 2, 2);",
+            "INSERT INTO books (code_identifier, isbn, release_date, title, subtitle, category_id, language_id, publisher_id) VALUES ('SE003', '1234567891013', '2100-01-01', 'Javascript Book', 'Why you shouldn''t learn it', 1, 3, 1);",
+            "INSERT INTO books_authors (book_id, person_id) VALUES (1, 1), (1, 2), (2, 2), (3, 1), (3, 2);"
+        ];
```

Comment:
Do you want to keep all the queries here? You might have other tests which could also use them. Maybe better put them in a separate module?

Reply:
Initially planned to put that into a separate PR, but I can also refactor it now

Reply:
I meant the source file, not the PR.

Reply:
Me too. "Initially I wanted to move the util functionality into a separate file in a separate PR".